src/SecretsFile.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module SecretsFile where
+
+import Control.Applicative.Combinators (some, option, optional, sepBy1)
+import Control.Exception (try)
+import Control.Monad.Except (MonadError, MonadIO, liftEither, liftIO)
+import Data.Char (toUpper, isSpace)
+import Data.Functor (void)
+import Data.List (intercalate)
+import Data.Void (Void)
+import qualified Text.Megaparsec as MP
+import qualified Text.Megaparsec.Char as MPC
+import qualified Text.Megaparsec.Char.Lexer as MPL
+
+data Secret = Secret
+  { sMount :: String
+  , sPath :: String
+  , sKey :: String
+  , sVarName :: String
+  } deriving (Eq, Show)
+
+data SFVersion
+  = V1
+  | V2
+  deriving (Show)
+
+type Parser = MP.Parsec Void String
+
+data SFError = IOError | ParseError (MP.ParseError (MP.Token String) Void)
+
+instance Show SFError where
+  show sfErr =
+    case sfErr of
+      IOError -> "could not read file"
+      ParseError pe -> MP.parseErrorPretty pe
+
+-- | Helper for ExceptT stuff that we use in app/Main.hs
+readSecretList :: (MonadError SFError m, MonadIO m) => FilePath -> m [Secret]
+readSecretList fp = liftEither =<< (liftIO $ readSecretsFile fp)
+
+-- | Read a lit of secrets from a file
+readSecretsFile :: FilePath -> IO (Either SFError [Secret])
+readSecretsFile fp = do
+  contents <- safeReadFile fp
+  case contents of
+    Just c -> do
+      let parseResult = parseSecretsFile fp c
+      case parseResult of
+        Right res -> pure $ Right res
+        Left err -> pure $ Left (ParseError err)
+    Nothing -> pure $ Left IOError
+
+-- | Read a file, catching all IOError exceptions.
+safeReadFile :: FilePath -> IO (Maybe String)
+safeReadFile fp = do
+  (contentsOrErr :: Either IOError String) <- (try . readFile) fp
+  case contentsOrErr of
+    Right contents -> pure $ Just contents
+    Left _ -> pure $ Nothing
+
+-- | Parse a String as a SecretsFile.
+parseSecretsFile :: FilePath -> String -> Either (MP.ParseError (MP.Token String) Void) [Secret]
+parseSecretsFile = MP.parse secretsFileP
+
+-- | SpaceConsumer parser, which is responsible for stripping all whitespace.
+--
+-- Sometimes, we require explicit newlines, therefore, we don't handle those
+-- here.
+whitespace :: Parser ()
+whitespace = MPL.space whitespaceChars lineComment blockComment
+  where
+    whitespaceChars = void $ MP.takeWhile1P (Just "whitespace") (\c -> isSpace c && c /= '\n')
```

Comment:
What does the `void` throw away here?

Reply:
It changes the type from `Parser String` to `Parser ()`. Here, that's the whitespace

Reply:
You are only parsing `isSpace` here. Isn't there an `isWhitespace` combinator already? Or would that also throw away newlines then? It would be nice to throw away *all* whitespace (including all the crazy unicode chars) except for newlines, I'd say.

Reply:
The builtin ones in Megaparsec indeed all get rid of the newlines, that's why we need our own here.

> It would be nice to throw away all whitespace (including all the crazy unicode chars) except for newlines, I'd say.

That's indeed what this does. `isSpace` comes from `Data.Char` and gives us all that for free :) I'll add a comment